src/tls.py

```diff
@@ -87,9 +131,93 @@ def _tls_relation_broken(self, _) -> None:
+            logger.error(msg)
+            self.charm.unit.status = BlockedStatus(msg)
+            return
+
+        self.charm.app_peer_data.update({"mtls": "enabled"})
+        self.charm.unit.status = ActiveStatus()
+
+    def _trusted_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Generate a CSR so the tls-certificates operator works as expected."""
+        alias = self.generate_alias(app_name=event.app.name, relation_id=event.relation.id)
+        csr = (
+            generate_csr(
+                add_unique_id_to_subject_name=alias,
+                private_key=self.private_key.encode("utf-8"),
+                subject=self.charm.unit_peer_data.get("private-address", ""),
+                **self._sans,
+            )
+            .decode()
+            .strip()
+        )
+
+        csr_dict = [{"certificate_signing_request": csr}]
+        event.relation.data[self.model.unit]["certificate_signing_requests"] = json.dumps(csr_dict)
+
+    def _trusted_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Overrides the requirer logic of TLSInterface."""
+        relation = event.relation
+        relation_data = _load_relation_data(relation.data[relation.app])
+        provider_certificates = relation_data.get("certificates", [])
+
+        if not provider_certificates:
+            logger.warning("No certificates on provider side")
+            event.defer()
+            return
+
+        alias = self.generate_alias(relation.app.name, relation.id)
+        # NOTE: Relation should only be used with one set of certificates,
+        # hence using just the first item on the list.
+        content = (
+            provider_certificates[0]["certificate"]
+            if relation.name == TRUSTED_CERTIFICATES_RELATION
+            else provider_certificates[0]["ca"]
+        )
+        filename = (
+            f"{alias}_cert.pem"
+            if relation.name == TRUSTED_CERTIFICATES_RELATION
+            else f"{alias}_ca.pem"
+        )
+
+        safe_write_to_file(content=content, path=f"{SNAP_CONFIG_PATH}/{filename}")
+        self.import_cert(alias=f"{alias}", filename=filename)
+
+        self.charm.unit.status = ActiveStatus()
+
+    def _trusted_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle relation broken for a trusted certificate/ca relation."""
+        # All units will need to remove the cert from their truststore
+        relation = event.relation
+        alias = self.generate_alias(app_name=relation.app.name, relation_id=relation.id)
+        self.remove_cert(alias=alias)
+
+        # The leader will also handle removing the "mtls" flag if needed
+        if not self.charm.unit.is_leader():
+            return
+
+        # Get all relations, and remove the one being broken
+        all_relations = (
+            self.model.relations[TRUSTED_CA_RELATION]
+            + self.model.relations[TRUSTED_CERTIFICATES_RELATION]
```

Comment:
**todo:** IIRC, this will fail. If we have 2 relations, remove 1, this is fine. 

If we have 1, the other (which doesn't exist) will `KeyError` I think.

Reply:
Maybe set the `mtls` flag on some generic `config_changed`-esque event?

Reply:
hm, after testing I get empty list when the relation doesn't exist. So it should be fine.